src/virtio-serializer.h

```diff
@@ -0,0 +1,190 @@
+        if (!vq.read_desc_mem(a, desc_idx, offset, data_uchar, len)) {
+            return false;
+        }
+        data_uchar[len] = 0;
+        // Advance
+        offset += len;
+        return true;
+    }
+
+    template <typename T, typename... Rest>
+    bool unpack(T *pval, Rest... rest) {
+        if (!read_value(pval)) {
+            return false;
+        }
+        if constexpr (sizeof...(Rest) > 0) {
+            return unpack(rest...);
+        }
+        return true;
+    }
+};
+
+/// \brief Utility for packing formatted bytes into a Virtqueue buffer
+struct virtq_serializer {
+    i_device_state_access *a;
+    virtq &vq;
+    uint32_t queue_idx;
+    uint32_t desc_idx;
+    uint32_t offset;
+    uint32_t length;
+
+    explicit virtq_serializer(i_device_state_access *a, virtq &vq, uint32_t queue_idx, uint32_t desc_idx,
+        uint32_t offset = 0) :
+        a(a),
+        vq(vq),
+        queue_idx(queue_idx),
+        desc_idx(desc_idx),
+        offset(offset),
+        length(offset) {}
+    virtq_serializer() = delete;
+    ~virtq_serializer() = default;
+    virtq_serializer(const virtq_serializer &other) = delete;
+    virtq_serializer(virtq_serializer &&other) = delete;
+    virtq_serializer &operator=(const virtq_serializer &other) = delete;
+    virtq_serializer &operator=(virtq_serializer &&other) = delete;
+
+    bool write_bytes(const unsigned char *data, uint32_t data_len) {
+        if (!vq.write_desc_mem(a, desc_idx, offset, data, data_len)) {
+            return false;
+        }
+        // Advance
+        offset += data_len;
+        length = std::max(length, offset);
+        return true;
+    }
+
+    bool write_u16_string(const void *data, uint16_t data_len) {
+        // Write the string size
+        if (!write_value(&data_len)) {
+            return false;
+        }
+        // Write the string
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        if (!vq.write_desc_mem(a, desc_idx, offset, reinterpret_cast<const unsigned char *>(data), data_len)) {
+            return false;
+        }
+        // Advance
+        offset += data_len;
+        length = std::max(length, offset);
+        return true;
+    }
+
+    template <typename T>
+    bool write_value(const T *pval) {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        if (!vq.write_desc_mem(a, desc_idx, offset, reinterpret_cast<const unsigned char *>(pval), sizeof(T))) {
+            return false;
+        }
+        // Advance
+        offset += sizeof(T);
+        length = std::max(length, offset);
+        return true;
+    }
+
+    bool write_value(const char *pval) {
+        return write_u16_string(pval, strlen(pval));
+    }
+
+    template <int N>
+    bool write_value(const char pval[N]) {
+        return write_u16_string(pval, strnlen(pval, N));
+    }
+
+    template <typename T, typename... Rest>
+    bool pack(T *pval, Rest... rest) {
+        if (!write_value(pval)) {
```

Comment:
Fold expression?

Reply:
C++17 introduces fold expressions with the "comma" operator to simplify variadic templates. You almost never need recursion anymore.

Reply:
It even strengthened the expression evaluation order guarantees of C++ to make sure these expressions are well defined. A thing of beauty.

Reply:
Cool, I was not aware, I made use of it to simplify the code. I guess I am still stuck in C++11 :sweat_smile: